cc/keystone/middleware/lifesaver.py

```diff
@@ -145,12 +149,79 @@ def get_user(self, request):
 
+
+    def get_token(self, request: Request) -> None | str:
+        """
+        Tries to fetch token id from the request
+        :param request: the clients request
+        :return: str 'token_id'
+        """
+
+        # shortcut for version discovery request
+        if '/v3/' == request.path:
+            return None
+
+        token_id = None
+        try:
+            if '/v3/auth/tokens' == request.path and 'POST' == request.method:
+                body = request.json_body
+                auth = body.get('auth', {})
+                identity = auth.get('identity', {})
+                token = identity.get('token', {})
+                token_id = token.get('id', None)
+            elif '/v3/auth/tokens' == request.path and 'GET' == request.method:
+                if "X-Subject-Token" in request.headers.keys():
+                    token_id = request.headers.get("X-Subject-Token")
+        except JSONDecodeError as e:
+            self.logger.error("Could not decode JSON request: %s %s", request, e)
+        except Exception as e:
+            self.logger.error("Could not extract token from request: %s %s", request, e)
+
+        return token_id
+
+
     def process_request(self, request):
         return self.verify_request(request)
 
     def process_response(self, response, request=None):
         return self.verify_request(request, response)
 
+    def get_costs(self, status: int, validation_type: str) -> int:
+        cost = 0
+        if status >= 400:
+            # what penalty should be applied?
+            if validation_type == "User":
+                cost = self.utils.status_cost['default']
+                if str(status) in self.utils.status_cost:
+                    cost = self.utils.status_cost[str(status)]
+            elif validation_type == "Token":
+                cost = self.utils.token_cost['default']
+                if str(status) in self.utils.token_cost:
+                    cost = self.utils.token_cost[str(status)]
+        return cost
+
+    def calculate_score(self, request, response, item, item_score, validation_type="User"):
+        status = response.status_code
+        cost = self.get_costs(status, validation_type)
+        # deduct user credit ?
+        if int(cost) > 0:
+            # mark request as processed
+            request.environ['lifesaver'] = item
+            item_score.reduce(int(cost))
+            # update score metadata in case the configuration has changed
+            if item_score.credit != self.utils.credit:
+                item_score.credit = self.utils.credit
+            if item_score.refill_time != self.utils.refill_time:
+                item_score.refill_time = self.utils.refill_time
+            if item_score.refill_amount != self.utils.refill_amount:
+                item_score.refill_amount = self.utils.refill_amount
+
+            self.utils.set_score(item, item_score)
+            self.logger.info("%s has a remaining credit of %d - request %s %s returned %d" % (
```

Comment:
This would effectively store a token in logs. Are we sure that valid tokens - https://github.com/sapcc/keystone-extensions/pull/10/files#r2556797581 - will never be stored?